src/coreclr/jit/valuenum.cpp

```diff
@@ -3244,87 +3383,101 @@ bool ValueNumStore::VNEvalShouldFold(var_types typ, VNFunc func, ValueNum arg0VN
 ValueNum ValueNumStore::EvalUsingMathIdentity(var_types typ, VNFunc func, ValueNum arg0VN, ValueNum arg1VN)
 {
     ValueNum resultVN = NoVN; // set default result to unsuccessful
+    ValueNum ZeroVN;
+    ValueNum OneVN;
 
     if (typ == TYP_BYREF) // We don't want/need to optimize a zero byref
     {
         return resultVN; // return the unsuccessful value
     }
 
+    // (0 + x) == x
+    // (x + 0) == x
+    // This identity does not apply for floating point (when x == -0.0).
+    auto identityForAddition = [=]() -> ValueNum {
```

Comment:
I am not very happy to use lambdas here as it appears they lead to quite the bulky codegen. Alternative would be to use a "local struct" with a static method.